File: crm_employee/test/features/leads/domain/lead_contact_test.dart
import 'package:crm_employee/features/leads/domain/lead_contact.dart';
import 'package:flutter_test/flutter_test.dart';

// Issue #143. "Punya kontak" is email OR phone that is not blank. The
// blank-handling is not pedantry: crm_be trims an email and stores what is
// left, so a submission of "   " can come back as "" rather than null — a lead
// that a naive `lead.email != null` would call contactable.
void main() {
  group('leadHasContact', () {
    test('is true when either one is present, or both', () {
      expect(leadHasContact(email: 'budi@example.com', phone: null), isTrue);
      expect(leadHasContact(email: null, phone: '0812 3456 7890'), isTrue);
      expect(leadHasContact(email: 'budi@example.com', phone: '0812'), isTrue);
    });

    test('is false when neither exists — null or missing', () {
      expect(leadHasContact(email: null, phone: null), isFalse);
      expect(leadHasContact(), isFalse);
    });

    test('treats an empty or whitespace-only value as absent', () {
      expect(leadHasContact(email: '', phone: ''), isFalse);
      expect(leadHasContact(email: '   ', phone: '\t\n'), isFalse);
    });

    test('lets one real value carry a blank neighbour', () {
      expect(leadHasContact(email: '   ', phone: '0812'), isTrue);
      expect(leadHasContact(email: 'budi@example.com', phone: '  '), isTrue);
    });

    test(
      'counts a phone that will not normalise for WhatsApp — it can still be dialled',
      () {
        expect(leadHasContact(email: null, phone: 'ext 12'), isTrue);
      },
    );
  });

  // The call button and the sentence explaining why it is off must agree. They
  // used to disagree at the edges: `phone != null && phone.isNotEmpty` enabled
  // Telepon for "   " while nothing said there was no number.
  group('hasPhoneNumber', () {
    test('agrees with leadHasContact about what a blank phone is', () {
      expect(hasPhoneNumber('0812 3456 7890'), isTrue);
      expect(hasPhoneNumber(null), isFalse);
      expect(hasPhoneNumber(''), isFalse);
      expect(hasPhoneNumber('   '), isFalse);
    });
  });

  // Issue #149: the email action shows only for a real address, with the
  // same blank rule as the badge.
  group('hasEmailAddress', () {
    test('agrees with leadHasContact about what a blank email is', () {
      expect(hasEmailAddress('budi@example.com'), isTrue);
      expect(hasEmailAddress(null), isFalse);
      expect(hasEmailAddress(''), isFalse);
      expect(hasEmailAddress('   '), isFalse);
    });
  });

  // #173 — the bar under Telepon/WhatsApp never goes quiet (brief §11.3).
  group('callActionsNote', () {
    test('no number at all: both buttons are off, and it says so', () {
      expect(callActionsNote(phone: null, phoneE164: null), 'Lead ini belum punya nomor telepon.');
      expect(callActionsNote(phone: '   ', phoneE164: null), 'Lead ini belum punya nomor telepon.');
    });

    test('a number that is not international: only WhatsApp is off, and it says so', () {
      expect(callActionsNote(phone: '12345', phoneE164: null), 'Nomor ini tidak bisa dipakai untuk WhatsApp.');
    });

    test('both work: no sentence', () {
      expect(callActionsNote(phone: '0812-3456-7890', phoneE164: '+6281234567890'), isNull);
    });
  });
}

File: crm_dashboard/src/lib/lead-contact.test.ts
import { describe, expect, it } from "vitest";
import { hasContact, primaryContact, shouldConfirmNoContact } from "./lead-contact";

// Issue #143. "Punya kontak" is email OR phone that is not blank. The
// blank-handling is not pedantry: crm_be trims an email and stores what is
// left, so a submission of "   " can come back as "" rather than null — a
// lead that a naive `lead.email != null` would call contactable.

describe("hasContact", () => {
  it("is true when either one is present, or both", () => {
    expect(hasContact({ email: "budi@example.com", phone: null })).toBe(true);
    expect(hasContact({ email: null, phone: "0812 3456 7890" })).toBe(true);
    expect(hasContact({ email: "budi@example.com", phone: "0812" })).toBe(true);
  });

  it("is false when neither exists — null, undefined, or missing", () => {
    expect(hasContact({ email: null, phone: null })).toBe(false);
    expect(hasContact({})).toBe(false);
    expect(hasContact({ email: undefined, phone: undefined })).toBe(false);
  });

  it("treats an empty or whitespace-only value as absent", () => {
    expect(hasContact({ email: "", phone: "" })).toBe(false);
    expect(hasContact({ email: "   ", phone: "\t\n" })).toBe(false);
  });

  it("lets one real value carry a blank neighbour", () => {
    expect(hasContact({ email: "   ", phone: "0812" })).toBe(true);
    expect(hasContact({ email: "budi@example.com", phone: "  " })).toBe(true);
  });

  it("does not count company or notes — free text nobody can call", () => {
    const lead = { email: null, phone: null, company: "PT Maju Jaya", notes: "hubungi 0812 3456 7890" };
    expect(hasContact(lead)).toBe(false);
  });

  it("counts a phone that will not normalise for WhatsApp — it can still be dialled", () => {
    expect(hasContact({ email: null, phone: "ext 12" })).toBe(true);
  });
});

describe("shouldConfirmNoContact", () => {
  it("asks once when there is no contact and nobody has confirmed yet", () => {
    expect(shouldConfirmNoContact({ email: "", phone: "" }, false)).toBe(true);
    expect(shouldConfirmNoContact({ email: "  ", phone: null }, false)).toBe(true);
  });

  it("does not ask again once confirmed — 'Tetap simpan' must go through", () => {
    expect(shouldConfirmNoContact({ email: "", phone: "" }, true)).toBe(false);
  });

  it("never asks when there is a contact, confirmed or not", () => {
    expect(shouldConfirmNoContact({ email: "budi@example.com", phone: "" }, false)).toBe(false);
    expect(shouldConfirmNoContact({ email: "", phone: "0812" }, false)).toBe(false);
    expect(shouldConfirmNoContact({ email: "budi@example.com", phone: "" }, true)).toBe(false);
  });
});

describe("primaryContact", () => {
  it("prefers the phone, falls back to the email", () => {
    expect(primaryContact({ phone: "0812-3456-7890", email: "dewi@contoh.id" })).toBe("0812-3456-7890");
    expect(primaryContact({ phone: null, email: "sinar@contoh.id" })).toBe("sinar@contoh.id");
  });

  it("treats a whitespace-only phone as absent, same as hasContact", () => {
    expect(primaryContact({ phone: "   ", email: "a@contoh.id" })).toBe("a@contoh.id");
  });

  it("shows a dash when there is nothing to show", () => {
    expect(primaryContact({ phone: "", email: "  " })).toBe("—");
  });
});
